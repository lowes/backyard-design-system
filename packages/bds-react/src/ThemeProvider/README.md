# @lowes-tech/bds-react/ThemeProvider

## Namespaces

- [ThemeProvider](modules/ThemeProvider.md)

## Interfaces

- [BackyardTheme](interfaces/BackyardTheme.md)
- [ThemeProviderProps](interfaces/ThemeProviderProps.md)

## References

### default

Renames and re-exports [ThemeProvider](README.md#themeprovider)

## Type aliases

### BaseTheme

Ƭ **BaseTheme**: typeof `Theme`

## Variables

### ThemeConsumer

• **ThemeConsumer**: `React.Consumer`<[`BackyardTheme`](interfaces/BackyardTheme.md)\> = `StyledComponentsThemeConsumer`

ThemeConsumer same as 'styled-components'

___

### ThemeContext

• **ThemeContext**: `React.Context`<[`BackyardTheme`](interfaces/BackyardTheme.md)\> = `StyledComponentsThemeContext`

ThemeContext same as 'styled-components'

___

### ThemeProvider

• **ThemeProvider**: `BDSFunctional`<[`ThemeProviderProps`](interfaces/ThemeProviderProps.md)\>

Backyard React Theme Provider

 <ThemeProvider theme="light" font="fellix" someOtherContext="value">
     <Grid.Container>
         ...
             <Button />
         ...
     </Grid.Container>
 </ThemeProvider>

Custom Theme Provider wrapper to provide extra context and overrides for Backyard Themes.

## Functions

### dropShadow

▸ `Const` **dropShadow**(`shadow`, `defaultShadow?`): `string`

Converts box-shadow to drop-shadow, if necessary.

Use for `filter` css instead of `box-shadow`.

**`example`**
```
dropShadow(theme.shadows.shadow_04)
 => 'drop-shadow(0px 2px 1px rgba(0, 0, 0, 0.2)) 
         drop-shadow(0px 1px 1px rgba(0, 0, 0, 0.14)) 
         drop-shadow(0px 1px 3px rgba(0, 0, 0, 0.12))'
```

#### Parameters

| Name | Type | Default value | Description |
| :------ | :------ | :------ | :------ |
| `shadow` | `string` | `undefined` | box-shadow |
| `defaultShadow` | `string` | `'none'` | default shadow if none given |

#### Returns

`string`

___

### getShape

▸ `Const` **getShape**(`shapeProp`, `shapeContext`): ``"rounded"`` \| ``"squared"``

Get the shape or shape context in priority

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `shapeProp` | `ExtendedShapes` | shape value from props |
| `shapeContext` | ``"rounded"`` \| ``"squared"`` | shape value from context |

#### Returns

``"rounded"`` \| ``"squared"``

___

### splitShadows

▸ `Const` **splitShadows**(`shadows`, `defaultShadow`): `string`[]

Split the shadows in a given string

**`example`**
```
splitShadows('0px 2px 1px rgba(0, 0, 0, 0.2), 0px 1px 1px rgba(0, 0, 0, 0.14), 0px 1px 3px rgba(0, 0, 0, 0.12)')
 => [
     '0px 2px 1px rgba(0, 0, 0, 0.2)',
     '0px 1px 1px rgba(0, 0, 0, 0.14)',
     '0px 1px 3px rgba(0, 0, 0, 0.12)'
 ]
```

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `shadows` | `string` | shadow string from backyard |
| `defaultShadow` | `string` | default shadow if none given |

#### Returns

`string`[]

___

### useBackyardTheme

▸ `Const` **useBackyardTheme**(`props?`): [`BackyardTheme`](interfaces/BackyardTheme.md)

#### Parameters

| Name | Type |
| :------ | :------ |
| `props` | `UseBackyardThemeProps` |

#### Returns

[`BackyardTheme`](interfaces/BackyardTheme.md)

___

### validateBackyardTheme

▸ `Const` **validateBackyardTheme**(`theme`, `name`): `boolean`

#### Parameters

| Name | Type |
| :------ | :------ |
| `theme` | [`BackyardTheme`](interfaces/BackyardTheme.md) |
| `name` | `string` |

#### Returns

`boolean`
