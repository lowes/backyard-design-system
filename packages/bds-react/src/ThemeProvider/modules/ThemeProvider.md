# Namespace: ThemeProvider

## Variables

### bdsName

â€¢ **bdsName**: `string`
