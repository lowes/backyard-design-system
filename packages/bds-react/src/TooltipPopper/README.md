# @lowes-tech/bds-react/TooltipPopper

## Interfaces

- [TooltipPopperProps](interfaces/TooltipPopperProps.md)

## References

### default

Renames and re-exports [TooltipPopper](README.md#tooltippopper)

## Type aliases

### TooltipPopperRef

Ƭ **TooltipPopperRef**: `TooltipRef`

## Variables

### TooltipPopper

• **TooltipPopper**: `BDSForwardRef`<[`TooltipPopperProps`](interfaces/TooltipPopperProps.md)\>

Backyard React TooltipPopper

**`todo`** Comments

## Functions

### testReset

▸ **testReset**(): `void`

#### Returns

`void`
