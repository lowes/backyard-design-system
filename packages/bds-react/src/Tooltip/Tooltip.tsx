import * as React from 'react'
import classNames from 'classnames'

import type { BackyardBaseProps, BDSForwardRef, KeyOf } from '../utils/typings'
import { useBackyardTheme, getShape, dropShadow } from '../ThemeProvider'
import type { BackyardTheme } from '../ThemeProvider'

import TooltipWrapper from './styles/TooltipWrapper'

/**
 * Backyard React Tooltip
 *
 * @todo Comments
 */
const Tooltip: BDSForwardRef<TooltipProps> = React.forwardRef<TooltipRef, TooltipProps>(
    function Tooltip(
        {
            arrow = 'top',
            arrowProps,
            children,
            className,
            icon,
            invisible = false,
            shadow = 'shadow_04',
            title,
            variant = 'low_contrast',
            shape: shapeProp, // = 'rounded',
            width = 'auto',
            style,
            ...props
        },
        ref,
    ) {
        // Get Backyard Theme
        const theme = useBackyardTheme({
            validate: true,
            name: 'Tooltip',
        })

        // Backyard Context from Theme
        const { shape: shapeContext } = theme.context

        // Calculate shape
        const shape = getShape(shapeProp, shapeContext)

        // Tooltip CSS vars
        const tooltipCSSVars = React.useMemo(
            () => ({
                // Tooltip drop shadow
                '--style-tooltip-drop-shadow': dropShadow(theme.shadows[shadow]),
                // Tooltip width
                '--style-tooltip-width': theme.sizes[width] || width,
                // Tooltip font family
                '--style-tooltip-font-family': theme.font.family[theme.context.font],
                // Tooltip font weight
                '--style-tooltip-font-weight':
                    theme.context.font === 'roboto'
                        ? theme.font.weight.medium
                        : theme.font.weight.semibold,
            }),
            [theme.name, width, shadow],
        )

        /**
         * Layout:
         *  <div wrapper>
         *      <span icon />
         *      <span>
         *          <div title />
         *          <div subtitle />
         *      </span>
         *  </div>
         */
        return (
            <TooltipWrapper
                className={classNames(
                    'tooltip',
                    `variant--${variant}`,
                    `shape--${shape}`,
                    {
                        invisible,
                    },
                    className,
                )}
                style={{
                    ...tooltipCSSVars,
                    ...style,
                }}
                {...props}
                ref={ref}
            >
                <span className="tooltip-body">
                    <div className="tooltip-title">
                        {icon ? <span className="tooltip-icon">{icon}</span> : null}
                        {title}
                    </div>
                    <div className="tooltip-subtitle">{children}</div>
                </span>
                <div className="tooltip-background">
                    {arrow !== 'none' ? (
                        <span
                            className={classNames('tooltip-arrow', `placement--${arrow}`)}
                            {...arrowProps}
                        />
                    ) : null}
                </div>
            </TooltipWrapper>
        )
    },
)

type TooltipRef = HTMLDivElement

interface TooltipProps extends BackyardBaseProps<TooltipRef> {
    /**
     * Position of arrow from the center of the tooltip
     * Note: This is opposite of `TooltipPopper`'s `placement` prop, which positions the tooltip from the reference
     *  For example, a `bottom` placed `TooltipPopper` will have the `arrow` be on top
     *  [Some Element]
     *       /|\
     *    [Tooltip]
     */
    arrow?:
        | 'none'
        | 'top-start'
        | 'top'
        | 'top-end'
        | 'left'
        | 'right'
        | 'bottom-start'
        | 'bottom'
        | 'bottom-end'
    /**
     * Arrow Props to extend arrow with
     */
    arrowProps?: BackyardBaseProps<HTMLSpanElement>
    /**
     * Children of subtitle span
     */
    children?: React.ReactNode
    /**
     * DOM Class Name
     */
    className?: string
    /**
     * Leading icon of tooltip
     */
    icon?: React.ReactElement
    /**
     * Whether or not tooltip is invisible
     */
    invisible?: boolean
    /**
     * Shadow token to use, or none
     *
     * @default 'shadow_04'
     */
    shadow?: 'none' | KeyOf<BackyardTheme['shadows']> | 'string'
    /**
     * Title text of Tooltip
     */
    title?: string
    /**
     * Variant of Tooltip
     */
    variant?: 'low_contrast' | 'high_contrast'
    /**
     * Shape of the tooltip
     */
    shape?: 'rounded' | 'squared'
    /**
     * Width of Tooltip
     * Defaults to auto
     * Limited by a maximum width of 320px (Desktop) and 256px (Mobile)
     */
    width?: string
}

Tooltip.bdsName = 'Tooltip'

export { Tooltip }

export type { TooltipProps, TooltipRef }

export default Tooltip
