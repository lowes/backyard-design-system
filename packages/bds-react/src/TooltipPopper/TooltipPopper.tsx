import * as React from 'react'
import { usePopper } from 'react-popper'
import { Modifier as PopperJSModifier, Options as PopperJSOptions } from '@popperjs/core/lib/types'

import 'core-js/es/array/find'
import 'core-js/es/promise'
import 'core-js/es/object/assign'

import useControlled from '../utils/hooks/useControlled'
import useRandomId from '../utils/hooks/useRandomId'
import useForkRef from '../utils/hooks/useForkRef'
import { usePopupState, bindHover, bindPopper } from '../utils/hooks/usePopupState'
import persistSyntheticEvent from '../utils/functions/persistSyntheticEvent'
import type { BDSForwardRef } from '../utils/typings/BDSComponentProps'
import { useBackyardTheme } from '../ThemeProvider'
import Tooltip from '../Tooltip'
import type { TooltipProps, TooltipRef } from '../Tooltip'
import Portal from '../Portal'
import type { PortalProps } from '../Portal'



let hysteresisOpen = false
let hysteresisTimer = null

export function testReset() {
    hysteresisOpen = false

    clearTimeout(hysteresisTimer)
}

const flipArrow = (placement) =>
    ({
        'bottom-end': 'top-start',
        'bottom-start': 'top-end',
        bottom: 'top',
        'left-end': 'right-start',
        'left-start': 'right-end',
        left: 'right',
        'right-end': 'left-start',
        'right-start': 'left-end',
        right: 'left',
        'top-end': 'bottom-start',
        'top-start': 'bottom-end',
        top: 'bottom',
    }[placement])

/**
 * Backyard React TooltipPopper
 *
 * @todo Comments
 */
const TooltipPopper: BDSForwardRef<TooltipPopperProps> = React.forwardRef<
    TooltipPopperRef,
    TooltipPopperProps
>(function TooltipPopper(
    {
        children,
        className,
        container,
        disablePortal = false,
        enterDelay = 200,
        enterNextDelay = 0,
        id: idProp,
        keepMounted = false,
        leaveDelay = 200,
        onClose,
        onOpen,
        open: openProp,
        defaultOpen = false,
        placement = 'bottom',
        popperModifiers = [],
        popperOptions = {},
        tooltip,
        title,
        ...props
    },
    ref,
) {
    const theme = useBackyardTheme({ validate: true, name: 'TooltipPoppper' })

    const [referenceElement, setReferenceElement] = React.useState(null)
    const [popperElement, setPopperElement] = React.useState(null)
    const [arrowElement, setArrowElement] = React.useState(null)
    const popperRef = useForkRef(setPopperElement, ref)

    const enterTimer = React.useRef<NodeJS.Timeout>()
    const leaveTimer = React.useRef<NodeJS.Timeout>()

    const id = useRandomId(idProp)

    const [open, setOpen] = useControlled({
        controlled: openProp,
        default: defaultOpen,
        name: `Backyard->TooltipPopper->[id="${id}"]`,
    })

    const popupState = usePopupState({ variant: 'popper', popupId: id })

    const { open: hovering, ...popperProps } = bindPopper(popupState)
    const { onMouseEnter, onMouseLeave, ...hoverProps } = bindHover(popupState)

    React.useEffect(() => {
        return () => {
            clearTimeout(enterTimer.current)
            clearTimeout(leaveTimer.current)
        }
    }, [])

    const handleOpen = (event) => {
        clearTimeout(hysteresisTimer)
        hysteresisOpen = true

        setOpen(true)

        if (onOpen) {
            onOpen(event)
        }
    }

    const handleClose = (event) => {
        clearTimeout(hysteresisTimer)

        hysteresisTimer = setTimeout(() => {
            hysteresisOpen = false
        }, 800 + leaveDelay)

        setOpen(false)

        if (onClose) {
            onClose(event)
        }
    }

    const handleEnter = (event) => {
        onMouseEnter(event)

        clearTimeout(enterTimer.current)
        clearTimeout(leaveTimer.current)

        if (enterDelay || (hysteresisOpen && enterNextDelay)) {
            persistSyntheticEvent(event)

            enterTimer.current = setTimeout(
                () => {
                    handleOpen(event)
                },
                hysteresisOpen ? enterNextDelay : enterDelay,
            )
        } else {
            handleOpen(event)
        }
    }

    const handleLeave = (event) => {
        onMouseLeave(event)

        clearTimeout(enterTimer.current)
        clearTimeout(leaveTimer.current)

        persistSyntheticEvent(event)

        leaveTimer.current = setTimeout(() => {
            handleClose(event)
        }, leaveDelay)
    }

    const { styles, attributes } = usePopper(referenceElement, popperElement, {
        modifiers: [
            { name: 'arrow', options: { element: arrowElement, padding: 8 } },
            { name: 'offset', options: { offset: [0, 16] } },
            { name: 'eventListeners', options: { scroll: open, resize: open } },
            ...popperModifiers,
        ],
        placement,
        ...popperOptions,
    })

    const place = attributes.popper ? attributes.popper['data-popper-placement'] : placement

    return (
        <>
            {React.cloneElement(children, {
                ...hoverProps,
                onMouseEnter: handleEnter,
                onMouseLeave: handleLeave,
                ref: setReferenceElement,
            })}
            {open || keepMounted ? (
                <Portal disablePortal={disablePortal} container={container}>
                    <Tooltip
                        id={id}
                        {...tooltip.props}
                        {...attributes.popper}
                        {...popperProps}
                        ref={popperRef}
                        arrowProps={{
                            ref: setArrowElement,
                            style: styles.arrow,
                        }}
                        arrow={flipArrow(place)}
                        style={{
                            ...styles.popper,
                            zIndex: theme.zIndex.tooltip,
                            opacity: openProp || hovering ? 1 : 0,
                        }}
                        {...props}
                    />
                </Portal>
            ) : null}
        </>
    )
})

type TooltipPopperRef = TooltipRef

type TooltipPopperEvent = React.MouseEvent

type TooltipPopperOverrideProps = 'title'

interface TooltipPopperProps extends Omit<TooltipProps, TooltipPopperOverrideProps> {
    /**
     * Tooltip reference element.
     */
    children: React.ReactElement<any, any>
    /**
     * DOM Class Name
     */
    className?: string
    /**
     * Portal Container
     */
    container?: PortalProps['container']
    /**
     * Whether or not the tooltip is open by default
     */
    defaultOpen?: boolean
    /**
     * Whether or not portal will be disabled
     */
    disablePortal?: PortalProps['disablePortal']
    /**
     * The number of milliseconds to wait before showing the tooltip.
     * This prop won't impact the enter touch delay (`enterTouchDelay`).
     */
    enterDelay?: number
    /**
     * The number of milliseconds to wait before showing the tooltip when one was already recently opened.
     */
    enterNextDelay?: number
    /**
     * This prop is used to help implement the accessibility logic.
     * If you don't provide this prop. It falls back to a randomly generated id.
     */
    id?: string
    /**
     * Whether or not the tooltip/portal will remain mounted when closed
     * This is useful for SSR/SEO situations where you want the Tooltip to remain on the page when closed
     */
    keepMounted?: boolean
    /**
     * The number of milliseconds to wait before hiding the tooltip.
     * This prop won't impact the leave touch delay (`leaveTouchDelay`).
     */
    leaveDelay?: number
    /**
     * Callback fired when the component requests to be closed.
     *
     * @param {Event} event The event source of the callback.
     */
    onClose?: (event: TooltipPopperEvent) => void
    /**
     * Callback fired when the component requests to be open.
     *
     * @param {object} event The event source of the callback.
     */
    onOpen?: (event: TooltipPopperEvent) => void
    /**
     * If `true`, the tooltip is shown.
     */
    open?: boolean
    /**
     * Position of the tooltip from the reference element (child)
     * Note: This is opposite of `Tooltip`'s `arrow` prop, which positions the tooltip from the reference
     *  For example, a `bottom` placed `TooltipPopper` will have the `arrow` be on top
     *  [Some Element]
     *       /|\
     *    [Tooltip]
     */
    placement?:
        | 'bottom-end'
        | 'bottom-start'
        | 'bottom'
        | 'left-end'
        | 'left-start'
        | 'left'
        | 'right-end'
        | 'right-start'
        | 'right'
        | 'top-end'
        | 'top-start'
        | 'top'
    /**
     * PopperJS Modifiers to extend/override default modifiers in `TooltipPopper`
     */
    popperModifiers?: PopperJSModifier<string, object>[]
    /**
     * PopperJS Options to extends/override default options in `TooltipOptions`
     */
    popperOptions?: PopperJSOptions
    /**
     * Tooltip title. Zero-length titles string are never displayed.
     */
    title?: React.ReactNode
    /**
     * Tooltip to handle popper events
     */
    tooltip: React.ReactElement<TooltipProps>
}

TooltipPopper.bdsName = 'TooltipPopper'

export { TooltipPopper }

export { TooltipPopperProps, TooltipPopperRef }

export default TooltipPopper
