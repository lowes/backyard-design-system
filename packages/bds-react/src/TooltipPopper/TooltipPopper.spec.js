import * as React from 'react'
import { render } from '../../test-utils'
import { ThemeProvider } from '../ThemeProvider'
import Tooltip from '../Tooltip'
import TooltipPopper from './TooltipPopper'
import { act } from 'react-dom/test-utils'

describe('TooltipPopper Snapshots', () => {
    test('TooltipPopper default snapshot', () => {
        const { asFragment } = render(
            <TooltipPopper
                id="test-popper"
                tooltip={(
                    <Tooltip>
                        Subtitle
                    </Tooltip>
                )}
            >
                <span>Child</span>
            </TooltipPopper>
        )

        expect(asFragment()).toMatchSnapshot()
    })


    test('TooltipPopper keepMounteded default snapshot', async () => {
        jest.useFakeTimers()

        const { asFragment } = render(
            <TooltipPopper
                id="test-popper"
                tooltip={(
                    <Tooltip>
                        Subtitle
                    </Tooltip>
                )}
                open
            >
                <span>Child</span>
            </TooltipPopper>
        )

        await act(async () => {
            jest.runAllTimers()
        })

        expect(asFragment()).toMatchSnapshot()
    })

    describe('themes', () => {
        const themes = ['light', 'dark']

        themes.forEach((theme) => {
            test(`TooltipPopper ${theme} theme snapshot`, async () => {
                jest.useFakeTimers()

                const { asFragment } = render(
                    <ThemeProvider theme={theme}>
                        <React.Fragment>
                            <TooltipPopper
                                id="test-popper"
                                tooltip={(
                                    <Tooltip>
                                        Subtitle
                                    </Tooltip>
                                )}
                                open
                            >
                                <span>Child</span>
                            </TooltipPopper>
                        </React.Fragment>
                    </ThemeProvider>
                )

                await act(async () => {
                    jest.runAllTimers()
                })

                expect(asFragment()).toMatchSnapshot()
            })
        })
    })

    describe('prop: placement', () => {
        const placements = [
            'top-start',
            'top',
            'top-end',
            'left',
            'right',
            'bottom-start',
            'bottom',
            'bottom-end'
        ]

        placements.forEach((placement) => {
            test(`TooltipPopper ${placement} placement snapshot`, async () => {
                jest.useFakeTimers()

                const { asFragment } = render(
                    <TooltipPopper
                        id="test-popper"
                        tooltip={(
                            <Tooltip>
                                Subtitle
                            </Tooltip>
                        )}
                        open
                        placement={placement}
                    >
                        <span>Child</span>
                    </TooltipPopper>
                )

                await act(async () => {
                    jest.runAllTimers()
                })

                expect(asFragment()).toMatchSnapshot()
            })
        })
    })

    describe('prop: disablePortal', () => {
        test(`TooltipPopper disabledPortal snapshot`, async () => {
            jest.useFakeTimers()

            const { asFragment } = render(
                <TooltipPopper
                    id="test-popper"
                    tooltip={(
                        <Tooltip>
                            Subtitle
                        </Tooltip>
                    )}
                    open
                    disablePortal
                >
                    <span>Child</span>
                </TooltipPopper>
            )

            await act(async () => {
                jest.runAllTimers()
            })

            expect(asFragment()).toMatchSnapshot()
        })
    })

    describe('prop: keepMounted', () => {
        test(`TooltipPopper keepMounted snapshot`, async () => {
            jest.useFakeTimers()

            const { asFragment } = render(
                <TooltipPopper
                    id="test-popper"
                    tooltip={(
                        <Tooltip>
                            Subtitle
                        </Tooltip>
                    )}
                    open
                >
                    <span>Child</span>
                </TooltipPopper>
            )

            await act(async () => {
                jest.runAllTimers()
            })

            expect(asFragment()).toMatchSnapshot()
        })
    })
})
