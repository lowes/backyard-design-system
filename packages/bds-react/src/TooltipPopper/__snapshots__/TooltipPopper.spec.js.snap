// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TooltipPopper Snapshots TooltipPopper default snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots TooltipPopper keepMounteded default snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots prop: disablePortal TooltipPopper disabledPortal snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
  .c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: var(--style-tooltip-width);
  max-width: calc(var(--bds-sizes-size-128) * 2);
  z-index: var(--bds-z-index-tooltip);
  padding: var(--bds-sizes-size-12) var(--bds-sizes-size-16);
  -webkit-transition: opacity 0.2s ease-in-out,top 0s,left 0s,-webkit-transform 0s;
  -webkit-transition: opacity 0.2s ease-in-out,top 0s,left 0s,transform 0s;
  transition: opacity 0.2s ease-in-out,top 0s,left 0s,transform 0s;
}

.c0.shape--rounded .tooltip-background {
  border-top-left-radius: var(--bds-border-radius-lg);
  border-top-right-radius: var(--bds-border-radius-lg);
  border-bottom-left-radius: var(--bds-border-radius-lg);
  border-bottom-right-radius: var(--bds-border-radius-lg);
}

.c0.shape--squared .tooltip-background {
  border-radius: 0;
}

.c0.invisible {
  display: none;
}

.c0 .tooltip-background {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  -webkit-filter: var(--style-tooltip-drop-shadow);
  filter: var(--style-tooltip-drop-shadow);
}

.c0 .tooltip-icon {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: var(--bds-sizes-size-2);
  z-index: 1;
}

.c0 .tooltip-body {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  z-index: 1;
}

.c0 .tooltip-body .tooltip-title {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: var(--style-tooltip-font-family);
  font-weight: var(--style-tooltip-font-weight);
  font-size: var(--bds-sizes-size-14);
  line-height: var(--bds-sizes-size-24);
}

.c0 .tooltip-body .tooltip-subtitle {
  display: block;
  font-family: var(--style-tooltip-font-family);
  font-weight: var(--bds-font-weight-regular);
  font-size: var(--bds-sizes-size-12);
  line-height: var(--bds-sizes-size-16);
  padding-bottom: var(--bds-sizes-size-1);
  padding-left: var(--bds-sizes-size-1);
}

.c0 .tooltip-arrow {
  -webkit-transition: opacity 0.2s ease-in-out,top 0s,left 0s,-webkit-transform 0s;
  -webkit-transition: opacity 0.2s ease-in-out,top 0s,left 0s,transform 0s;
  transition: opacity 0.2s ease-in-out,top 0s,left 0s,transform 0s;
}

.c0 .tooltip-arrow,
.c0 .tooltip-arrow::before {
  position: absolute;
  width: var(--bds-sizes-size-14);
  height: var(--bds-sizes-size-14);
  z-index: var(--bds-z-index-tooltip);
}

.c0 .tooltip-arrow::before {
  content: '';
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c0 .tooltip-arrow.placement--top,
.c0 .tooltip-arrow.placement--top-start,
.c0 .tooltip-arrow.placement--top-end {
  top: calc(-1 * var(--bds-sizes-size-7));
}

.c0 .tooltip-arrow.placement--top-start {
  left: var(--bds-sizes-size-16);
}

.c0 .tooltip-arrow.placement--top {
  left: calc(50% - var(--bds-sizes-size-6));
}

.c0 .tooltip-arrow.placement--top-end {
  left: calc(100% - var(--bds-sizes-size-32));
}

.c0 .tooltip-arrow.placement--left,
.c0 .tooltip-arrow.placement--left-start,
.c0 .tooltip-arrow.placement--left-end {
  left: calc(-1 * var(--bds-sizes-size-7));
}

.c0 .tooltip-arrow.placement--left-start {
  top: var(--bds-sizes-size-16);
}

.c0 .tooltip-arrow.placement--left {
  top: calc(50% - var(--bds-sizes-size-6));
}

.c0 .tooltip-arrow.placement--left-end {
  top: calc(100% - var(--bds-sizes-size-32));
}

.c0 .tooltip-arrow.placement--right,
.c0 .tooltip-arrow.placement--right-start,
.c0 .tooltip-arrow.placement--right-end {
  right: calc(-1 * var(--bds-sizes-size-7));
}

.c0 .tooltip-arrow.placement--right-start {
  top: var(--bds-sizes-size-16);
}

.c0 .tooltip-arrow.placement--right {
  top: calc(50% - var(--bds-sizes-size-6));
}

.c0 .tooltip-arrow.placement--right-end {
  top: calc(100% - var(--bds-sizes-size-32));
}

.c0 .tooltip-arrow.placement--bottom,
.c0 .tooltip-arrow.placement--bottom-start,
.c0 .tooltip-arrow.placement--bottom-end {
  bottom: calc(-1 * var(--bds-sizes-size-6));
}

.c0 .tooltip-arrow.placement--bottom-start {
  left: var(--bds-sizes-size-16);
}

.c0 .tooltip-arrow.placement--bottom {
  left: calc(50% - var(--bds-sizes-size-6));
}

.c0 .tooltip-arrow.placement--bottom-end {
  left: calc(100% - var(--bds-sizes-size-32));
}

.c0.variant--low_contrast .tooltip-background {
  background: var(--bds-color-elevation-04);
}

.c0.variant--low_contrast .tooltip-icon svg path {
  fill: var(--bds-color-icon-primary);
}

.c0.variant--low_contrast .tooltip-title {
  color: var(--bds-color-text-primary);
}

.c0.variant--low_contrast .tooltip-subtitle {
  color: var(--bds-color-text-primary);
}

.c0.variant--low_contrast .tooltip-arrow::before {
  background: var(--bds-color-elevation-04);
}

.c0.variant--high_contrast .tooltip-background {
  background: var(--bds-color-surface-default-inverse);
}

.c0.variant--high_contrast .tooltip-icon svg path {
  fill: var(--bds-color-icon-primary-inverse);
}

.c0.variant--high_contrast .tooltip-title {
  color: var(--bds-color-text-primary-inverse);
}

.c0.variant--high_contrast .tooltip-subtitle {
  color: var(--bds-color-text-primary-inverse);
}

.c0.variant--high_contrast .tooltip-arrow::before {
  background: var(--bds-color-surface-default-inverse);
}

<div
    class="c0 tooltip variant--low_contrast shape--rounded"
    data-popper-escaped="true"
    data-popper-placement="bottom"
    data-popper-reference-hidden="true"
    id="test-popper"
    style="--style-tooltip-drop-shadow: drop-shadow(0px 5px 5px rgba(0, 0, 0, 0.2)) drop-shadow(0px 8px 10px rgba(0, 0, 0, 0.14)) drop-shadow(0px 3px 14px rgba(0, 0, 0, 0.12)); --style-tooltip-width: auto; --style-tooltip-font-family: Roboto, Helvetica, Arial, sans-serif; --style-tooltip-font-weight: 500; position: absolute; left: 0px; top: 0px; z-index: 1500; opacity: 1; transform: translate(0px, 16px);"
  >
    <span
      class="tooltip-body"
    >
      <div
        class="tooltip-title"
      />
      <div
        class="tooltip-subtitle"
      >
        Subtitle
      </div>
    </span>
    <div
      class="tooltip-background"
    >
      <span
        class="tooltip-arrow placement--top"
        style="position: absolute; left: 0px; transform: translate(8px, 0px);"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots prop: keepMounted TooltipPopper keepMounted snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots prop: placement TooltipPopper bottom placement snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots prop: placement TooltipPopper bottom-end placement snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots prop: placement TooltipPopper bottom-start placement snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots prop: placement TooltipPopper left placement snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots prop: placement TooltipPopper right placement snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots prop: placement TooltipPopper top placement snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots prop: placement TooltipPopper top-end placement snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots prop: placement TooltipPopper top-start placement snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots themes TooltipPopper dark theme snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;

exports[`TooltipPopper Snapshots themes TooltipPopper light theme snapshot 1`] = `
<DocumentFragment>
  <span>
    Child
  </span>
</DocumentFragment>
`;
